lib_eio/buf_read.ml

```diff
@@ -143,6 +143,98 @@ let as_flow t =
 let get t i =
   Bigarray.Array1.get t.buf (t.pos + i)
 
+module BE = struct
+  let uint16 t = 
+    ensure t 2;
+    let data = Bigstringaf.get_int16_be t.buf t.pos in
+    consume t 2;
+    data
+
+  let uint32 t = 
+    ensure t 4;
+    let data = Bigstringaf.get_int32_be t.buf t.pos in
+    consume t 4;
+    data
+
+  let uint48 t = 
+    ensure t 6;
+    let upper_16 = Bigstringaf.get_int16_be t.buf t.pos |> Int64.of_int in
+    let middle_16 = Bigstringaf.get_int16_be t.buf (t.pos + 2) |> Int64.of_int in
+    let lower_16 = Bigstringaf.get_int16_be t.buf (t.pos + 4) |> Int64.of_int in
+    consume t 6;
+    Int64.(
+      add 
+        (shift_left upper_16 32)
+      (add
+        (shift_left middle_16 16)
+        (lower_16))
+    )
+
+  let uint64 t =
+    ensure t 8;
+    let data = Bigstringaf.get_int64_be t.buf t.pos in
+    consume t 8;
+    data
+
+  let float t =
+    ensure t 4;
+    let data = Bigstringaf.unsafe_get_int32_be t.buf t.pos in
+    consume t 4;
+    Int32.float_of_bits data
+
+  let double t =
+    ensure t 8;
+    let data = Bigstringaf.unsafe_get_int64_be t.buf t.pos in
+    consume t 8;
+    Int64.float_of_bits data
+end
+
+module LE = struct
+  let uint16 t = 
+    ensure t 2;
+    let data = Bigstringaf.get_int16_le t.buf t.pos in
+    consume t 2;
+    data
+
+  let uint32 t = 
+    ensure t 4;
+    let data = Bigstringaf.get_int32_le t.buf t.pos in
+    consume t 4;
+    data
+
+  let uint48 t = 
+    ensure t 6;
+    let lower_16 = Bigstringaf.get_int16_le t.buf t.pos |> Int64.of_int in
+    let middle_16 = Bigstringaf.get_int16_le t.buf (t.pos + 2) |> Int64.of_int in
```

Comment:
This might be better (haven't benchmarked it though):
```suggestion
    let lower_32 = Bigstringaf.get_int32_le t.buf t.pos |> Int64.of_int |> Int64.log_and 0xffffffffL in
```
Then the total is just `Int64.(logor lower_32 (shift_left upper_16 32))`.

Reply:
I'm not sure this is correct for `-1L` since the of_int moves the sign bit such that the value is the same (but not the bits).
This was the correctness bug I was concerned about previously.

`logor` may be faster though than just adding.

Reply:
It doesn't move the sign bit, it just replicates it into the extra bits (which is correct for signed numbers but not for unsigned ones). The `logand` gets rid of the extra copies:
```
# Int64.of_int32 0xe7654321l |> Printf.sprintf "0x%Lx";;
- : string = "0xffffffffe7654321"
# Int64.of_int32 0xe7654321l |> Int64.logand 0xffffffffL |> Printf.sprintf "0x%Lx";;
- : string = "0xe7654321"
```

Reply:
Good to know! I've left the fuzz model using the 3 * 16 bit one for validation.